uqcsbot/scripts/leet.py

```diff
@@ -0,0 +1,167 @@
+from uqcsbot import bot, Command
+from http import HTTPStatus
+from uqcsbot.utils.command_utils import loading_status, UsageSyntaxException
+import json
+import requests
+import random
+from slackblocks import Attachment, SectionBlock
+
+
+LC_DIFFICULTY_MAP = ["easy", "medium", "hard"] # leetcode difficulty is 1,2,3, need to map
+HR_DS_API_LINK = ("https://www.hackerrank.com/rest/contests/master/tracks/" +
+    "data-structures/challenges?limit=200")
+HR_ALG_API_LINK = "https://www.hackerrank.com/rest/contests/master/tracks/algorithms/challenges?limit=200"
+LC_API_LINK = 'https://leetcode.com/api/problems/all/'
+
+EASY_COLOR_CODE = "#5db85b"
+MED_COLOR_CODE = "#f1ad4e"
+HARD_COLOR_CODE = "#d9534f"
+
+@bot.on_command('leet')
+@loading_status
+def handle_leet(command: Command) -> None:
+    """
+    `!leet [`easy` | `medium` | `hard`] - Retrieves a set of questions from online coding
+    websites, and posts in channel with a random question from this set. If a difficulty
+    is provided as an argument, the random question will be restricted to this level of 
+    challenge. Else, a random difficulty is generated to choose.
+    """
+
+    difficulty = ""
+    wasRandom = True # Used for output later
+
+    if command.has_arg():
+        if (command.arg not in {"easy", "medium", "hard"}):
+           bot.post_message(command.channel_id, "Usage: !leet [`easy` | `medium` | `hard`]")
+           return
+        else:
+            difficulty = command.arg.lower()
+            wasRandom = False
+
+    # Storage dict for the three categories, difficulty is key
+    questions = {"hard" : [], "medium" : [], "easy" : []}
+
+    # Go fetch questions from APIs
+    collect_questions(questions)
+
+    # No difficulty specified, randomly generate one
+    if (len(difficulty) == 0):
+        difficulty = random.choice(LC_DIFFICULTY_MAP) 
+    
+    
+    selected_question = select_question(difficulty, questions) # Get a random question
```

Comment:
if you obtain the difficulty first, you can filter out to the correct difficulty at the same time you're collecting questions

Reply:
As all API's have to be queried regardless, I don't think this makes too much of a difference.

Reply:
Implemented